### PR TITLE
Migrate to Flowzone

### PR DESCRIPTION
**Pull request by ab77**

### CODE REVIEW

**Comment by ab77** [comment]
One way to hand over state from compose tests to custom publishing pipeline, would be to just upload the source artefact after docker_tests have concluded (i.e.):

```
      # ... after docker-compose sut

      - name: Compress source
        run: tar -acvf ${{ runner.temp }}/docker.tgz .

      - name: Upload artifact
        uses: actions/upload-artifact@v3
        with:
          name: docker-${{ github.event.pull_request.head.sha || github.event.head_commit.id }}
          path: ${{ runner.temp }}/docker.tgz
          retention-days: 1
```

This approach will allow us to continue (miss)using docker-compose tests flow to not only run tests, but to also generate and extract assets for publishing into the source tree as this PR demonstrates. The subsequent custom publish steps would then pick up the assets from the source tree and publish them as GH releases, or whatever.

The other option I suppose, is to just do all of the build/test/publish work in the custom flow, like we've done for Etcher/CLI, but that will require recreating a lot of logic that already exists in the Docker flow and disabling the core Docker logic somehow.

I am leaning towards the first option. Thoughts @Page- @lmbarros @klutchell ?

**Comment by ab77** [comment]
@Page- @lmbarros here is the minimum set of changes I was able to come up with to move balena-engine to Flowzone and have it test, build and publish a GitHub release.

Unit tests are executed using docker-compose SUT pattern, reusing moby's DinD scripts. The build is done using docker buildx bake action, reusing moby's bake hcl. Everything is wrapped using Flowzone custom workflow.

We should review the skipped unit tests as well as give these binaries a solid test before merging. If that all works, we should probably consider consuming these releases in Jenkins balenaOS builds instead of rebuilding them there.

current run: https://github.com/balena-os/balena-engine/actions/runs/3471627690

upcoming draft release: https://github.com/balena-os/balena-engine/releases/tag/untagged-a3a09667183da3f1b5e3

cc @klutchell @alexgg @jakogut

File: Dockerfile
Comment (lmbarros) [inline Dockerfile:268]: Again, I'd keep this comment to not diverge from upstream. (This one is present even on Moby `master`.)

File: hack/make/.binary
Comment (lmbarros) [inline hack/make/.binary:91]: Yet another. This one I believe we can very safely remove until 22.06.

**Comment by ab77** [comment]
@nucleardreamer @vipulgupta2048 folks, https://www.balena.io/engine/ will most likely be invalidated by a new release here. Are one of you able to prepare a PR to match or let me know who else I can talk to about this please...

**Review by lmbarros — COMMENTED** [review]
I haven't looked at the skipped tests nor looked at the generated releases, but here's a first batch of comments.

The only thing I don't really like here is that many of the updates you made to the Dockerfile and scripts under `./hack` are coming from upstream's `master`, while balenaEngine currently follows their `20.10` branch -- which is still the stable branch. (Our last sync was with 20.10.12 and I have [this open PR](https://github.com/balena-os/balena-engine/pull/310) updating it to 20.10.17).

So, I'd like to keep as close as possible to 22.10 to avoid merge conflicts in future merges from upstream. I suspect many of the changes that are coming from master/22.06 aren't really necessary for the Flowzone migration.

I added comments to several of these points, but I didn't try to revert them and see what happens.

File: hack/make/.binary
Comment (ab77) [reply hack/make/.binary:70]: @lmbarros we have arm64/aarch64 release already published when Robert was still here. In any case these changes are to accommodate the bump in gcc version included with new Debian release.

File: hack/dind
Comment (ab77) [reply hack/dind:43]: @lmbarros might not be required, please feel free to experiment if you have the bandwidth.

File: docker-bake.hcl
Comment (ab77) [reply docker-bake.hcl:1]: @lmbarros this is required for the CI pipeline to run the builds. It's from moby/master and will be likely included in their next release (since they are also in the process of moving to GH actions).

File: Dockerfile
Comment (ab77) [reply Dockerfile:341]: @lmbarros This is also part of the upgrade to latest Debian. You are of course free to experiment, but I suggest we leave it as is.

File: Dockerfile
Comment (ab77) [reply Dockerfile:268]: @lmbarros yes, we can keep the comments.

File: hack/make/.binary
Comment (ab77) [reply hack/make/.binary:91]: @lmbarros I took the entire file and avoided hacking it. Feel free to remove manually if you like, but I would suggest to keep the file as is, since we don't "own it".

File: hack/make/.binary
Comment (ab77) [reply hack/make/.binary:70]: I built the release for all the architectures we've previously built for - what am I missing?

https://github.com/balena-os/balena-engine/releases/tag/v18.9.13

Comment (lmbarros) [reply hack/make/.binary:70]: > what am I missing?

Nothing! I was just clarifying that when I mentioned "only v5 and v7" I was referring to ARM32 specifically, and you were right about arm64. Again, keep in mind that my comments aren't really about the CI side of this PR (which is perfect AFAICT), but to the Engine maintenance perspective. My initial point was all about syncing to 20.10 versus 22.06.

**Comment by vipulgupta2048** [comment]
>  https://www.balena.io/engine/ will most likely be invalidated by a new release here. Are one of you able to prepare a PR to match or let me know who else I can talk to about this please...

cc: @iamsolankiamit Can help with the PR to balena-site so that the release is correct.

File: hack/make/.binary
Comment (ab77) [inline hack/make/.binary:52]: @lmbarros minimal change required to build `linux/arm/v6` (I have no idea how it was previously building without this).

File: hack/make/cross
Comment (ab77) [inline hack/make/cross:34]: @lmbarros this package is required to build `linux/386` (not sure if we even need it tbh); it wrecks ARM compiler config, so I am installing at runtime and building this target last.

**Comment by ab77** [comment]
> So, I'd like to keep as close as possible to 22.10 to avoid merge conflicts in future merges from upstream. I suspect many of the changes that are coming from master/22.06 aren't really necessary for the Flowzone migration.

@lmbarros I've done a quick sync with moby/20.10 and have managed to condense the number of changes to effectively 4 files. If you can reduce these even further, that would be awesome, but this is the best I was able to come up with.

**Comment by ab77** [comment]
> Unfortunately the current 20.10 Dockerfile, does not build in the GH Actions runner environment, even though it seems to work locally.

I got it to build with a few more external dependency bumps from 20.10 @lmbarros, over to you I guess...

**Comment by lmbarros** [comment]
So, @ab77 :

* I added a commit that moves the creation of the symlinks from the installer to the build process itself. This ensures that people downloading the archive directly from the GitHub releases will get the symlinks too.
* The tests being skipped are fine to be skipped.
* I tested some of the binaries (on a 32- and a 64-bit Raspberry Pi OS) and they are fine.

I believe at this point we are much better than we were before.

If you don't scream about this last commit I made, I'll merge this PR.

And thanks again for doing this! :pray:

**Comment by ab77** [comment]
> If you don't scream about this last commit I made, I'll merge this PR.

Ship it @lmbarros !

> And thanks again for doing this! 🙏

You are welcome, as always.

**Review by lmbarros — APPROVED** [review]
:rocket:

**Pull request merged**
